src/session.rs

```diff
@@ -0,0 +1,340 @@
+pub struct Cookie {
+    value: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    expires: Option<i64>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    path: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    secure: Option<bool>,
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+struct Content {
+    #[serde(rename = "__meta__")]
+    meta: Meta,
+    #[serde(skip_serializing)]
+    auth: Option<Auth>,
+    cookies: HashMap<String, Cookie>,
+    headers: HashMap<String, String>,
+}
+
+impl Content {
+    fn migrate(mut self) -> Self {
+        let auth = mem::take(&mut self.auth);
+        if let Some(Auth {
+            auth_type: Some(ref auth_type),
+            raw_auth: Some(ref raw_auth),
+        }) = auth
+        {
+            if auth_type.as_str() == "basic" {
+                self.headers
+                    .entry("authorization".into())
+                    .or_insert_with(|| format!("Basic {}", base64::encode(raw_auth)));
+            }
+        }
+        self.meta = Meta::default();
+        self
+    }
+}
+
+pub struct Session {
+    pub path: PathBuf,
+    pub read_only: bool,
+    content: Content,
+}
+
+impl Session {
+    pub fn load_session(url: &Url, mut name_or_path: OsString, read_only: bool) -> Result<Self> {
+        let path = if is_path(&name_or_path) {
+            PathBuf::from(name_or_path)
+        } else {
+            let mut path = config_dir()
+                .context("couldn't get config directory")?
+                .join::<PathBuf>(["xh", "sessions", &path_from_url(url)?].iter().collect());
+            name_or_path.push(".json");
+            path.push(name_or_path);
+            path
+        };
+
+        let content = match fs::read_to_string(&path) {
+            Ok(content) => serde_json::from_str::<Content>(&content)?.migrate(),
+            Err(err) if err.kind() == io::ErrorKind::NotFound => Content::default(),
+            Err(err) => return Err(err.into()),
+        };
+
+        Ok(Session {
+            path,
+            read_only,
+            content,
+        })
+    }
+
+    pub fn headers(&self) -> Result<HeaderMap> {
+        Ok(HeaderMap::try_from(&self.content.headers)?)
+    }
+
+    pub fn save_headers(&mut self, request_headers: &HeaderMap) -> Result<()> {
+        for (key, value) in request_headers.iter() {
+            let key = key.as_str();
+            if key != "cookie" && !key.starts_with("content-") && !key.starts_with("if-") {
+                self.content
+                    .headers
+                    .insert(key.into(), value.to_str()?.into());
+            }
+        }
+        Ok(())
+    }
+
+    pub fn save_auth(&mut self, request_headers: &HeaderMap) -> Result<()> {
+        if let Some(value) = request_headers.get(AUTHORIZATION) {
+            self.content
+                .headers
+                .insert("authorization".into(), value.to_str()?.into());
+        }
```

Comment:
Isn't this code as it is now redundant because of `save_headers()`?

Reply:
Not at the moment (that is, unless we stop using `RequestBuilder::basic_auth()`)

Reply:
`save_auth()` has been replaced with `save_basic_auth()` and `save_bearer_auth()`